allennlp/data/data_loaders/multi_process_data_loader.py

```diff
@@ -0,0 +1,304 @@
+                    if not self.lazy:
+                        self._instances.append(instance)  # type: ignore
+                    yield instance
+
+                self._join_workers(workers)
+
+    def index_with(self, vocab: Vocabulary) -> None:
+        self._vocab = vocab
+        if self._instances:
+            for instance in self._instances:
+                instance.index_fields(vocab)
+
+    def _start_instance_workers(self, queue: mp.JoinableQueue) -> List[mp.Process]:
+        workers: List[mp.Process] = []
+        for worker_id in range(self.num_workers):
+            worker = mp.Process(target=self._instance_worker, args=(worker_id, queue), daemon=True)
+            worker.start()
+            workers.append(worker)
+        return workers
+
+    def _join_workers(self, workers: List[mp.Process]) -> None:
+        for worker in workers:
+            # TODO: handle errors if any of the workers crash.
+            worker.join(1)
+            if worker.is_alive():
+                logger.warning("Worker is still alive, killing now")
+                worker.terminate()
+
+    def _gather_instances(self, queue: mp.JoinableQueue) -> Iterable[Instance]:
+        done_count: int = 0
+        while done_count < self.num_workers:
+            instances_chunk: Optional[List[Instance]]
+            for instances_chunk in iter(queue.get, None):
+                yield from instances_chunk
+                queue.task_done()
+            queue.task_done()
+            # Every time we encounter an empty list, thats means a worker has finished.
+            done_count += 1
+
+    def _instance_worker(self, worker_id: int, queue: mp.JoinableQueue) -> None:
+        self.reader._set_worker_info(WorkerInfo(self.num_workers, worker_id))
+
+        def index_fields(instance: Instance) -> Instance:
+            self.reader.apply_token_indexers(instance)
+            if self._vocab is not None:
+                instance.index_fields(self._vocab)
+            return instance
+
+        instances = (index_fields(instance) for instance in self.reader.read(self.data_path))
+
+        for instances_chunk in lazy_groups_of(instances, self._INSTANCE_CHUNK_SIZE):
+            queue.put(instances_chunk)
+
+        # Indicate to the consumer that this worker is finished.
+        queue.put(None)
+
+        # Wait for consumer to finish to avoid prematurely closing the queue.
+        queue.join()
+
+    def _instances_to_batches(self, instance_iterator: Iterable[Instance]) -> Iterator[TensorDict]:
+        instance_chunks: Iterable[List[Instance]]
+        if self.max_batches_in_memory is not None:
+            chunk_size = self.batch_size * self.max_batches_in_memory
+            instance_chunks = lazy_groups_of(instance_iterator, chunk_size)
+        else:
+            instance_chunks = [list(instance_iterator)]
+
+        for instances in instance_chunks:
+            if self.shuffle:
+                random.shuffle(instances)
```

Comment:
This is better than nothing, but it would be even better if we kept instances in a pool and then picked random subsets from that pool to make batches from. The way we have it now, early instances will never be shuffled together with late instances, but with a pool, that's an option.

It also occurs to me that the order of instances is now _always_ random, because it depends on the processing speed of the workers. In light of that, maybe we should set `self.shuffle` to `True` all the time and make it properly random all the time? It should also be called out in the documentation that you will not get exactly the same sequence twice.

Reply:
> The way we have it now, early instances will never be shuffled together with late instances....

That's true for lazy loading, but not for non-lazy loading.

Reply:
Right, but with the pool approach, you can make it not-true for lazy loading as well.